File: libs/sdk-bindings/bindings-android/lib/src/main/kotlin/breez_sdk_notification/job/LnurlPayInfo.kt
package breez_sdk_notification.job

import android.content.Context
import breez_sdk.BlockingBreezServices
import breez_sdk.OpenChannelFeeRequest
import breez_sdk_notification.Constants.DEFAULT_LNURL_PAY_INFO_NOTIFICATION_TITLE
import breez_sdk_notification.Constants.DEFAULT_LNURL_PAY_METADATA_PLAIN_TEXT
import breez_sdk_notification.Constants.DEFAULT_LNURL_PAY_NOTIFICATION_FAILURE_TITLE
import breez_sdk_notification.Constants.LNURL_PAY_INFO_NOTIFICATION_TITLE
import breez_sdk_notification.Constants.LNURL_PAY_METADATA_PLAIN_TEXT
import breez_sdk_notification.Constants.LNURL_PAY_NOTIFICATION_FAILURE_TITLE
import breez_sdk_notification.Constants.NOTIFICATION_CHANNEL_LNURL_PAY
import breez_sdk_notification.NotificationHelper.Companion.notifyChannel
import breez_sdk_notification.ResourceHelper.Companion.getString
import breez_sdk_notification.SdkForegroundService
import breez_sdk_notification.ServiceConfig
import breez_sdk_notification.ServiceLogger
import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

@Serializable
data class LnurlInfoRequest(
    @SerialName("callback_url") val callbackURL: String,
    @SerialName("reply_url") val replyURL: String,
)

// Serialize the response according to to LUD-06 payRequest base specification:
// https://github.com/lnurl/luds/blob/luds/06.md
@Serializable
data class LnurlPayInfoResponse(
    val callback: String,
    val maxSendable: ULong,
    val minSendable: ULong,
    val metadata: String,
    val tag: String,
)

class LnurlPayInfoJob(
    private val context: Context,
    private val fgService: SdkForegroundService,
    private val payload: String,
    private val logger: ServiceLogger,
    private val config: ServiceConfig,
) : LnurlPayJob {
    companion object {
        private const val TAG = "LnurlPayInfoJob"
    }

    override fun start(breezSDK: BlockingBreezServices) {
        var request: LnurlInfoRequest? = null
        try {
            request = Json.decodeFromString(LnurlInfoRequest.serializer(), payload)
            // Get the fee parameters offered by the LSP for opening a new channel
            val ofp =
                breezSDK.openChannelFee(OpenChannelFeeRequest(amountMsat = null)).feeParams
            // Calculate the maximum receivable amount within the fee limit (in millisatoshis)
            val feeLimitMsat: ULong = config.channelFeeLimitMsat
            val nodeInfo = breezSDK.nodeInfo()
            val proportionalPercent = ((ofp.proportional.toDouble() / 1000000.0))
            // Treat fee limit feature as disabled when it's set to 0
            val maxReceivableMsatFeeLimit = if (proportionalPercent != 0.0 && feeLimitMsat != 0UL) {
                minOf(
                    nodeInfo.maxReceivableMsat,
                    (feeLimitMsat.toDouble() / proportionalPercent).toULong()
                )
            } else {
                nodeInfo.maxReceivableMsat
            }
            // Calculate the maximum sendable amount (in millisatoshis)
            val maxSendable = maxOf(
                nodeInfo.inboundLiquidityMsats,
                maxReceivableMsatFeeLimit,
            )
            // Get the minimum sendable amount (in millisatoshis), can not be less than 1 or more than maxSendable
            val minSendable: ULong =
                if (nodeInfo.inboundLiquidityMsats < 1000UL) ofp.minMsat else 1000UL
            if (minSendable > maxSendable) {
                throw Exception("Minimum sendable amount can't be greater than maximum sendable amount.")
            }
            val plainTextMetadata = getString(
                context,
                LNURL_PAY_METADATA_PLAIN_TEXT,
                DEFAULT_LNURL_PAY_METADATA_PLAIN_TEXT
            )
            val response =
                LnurlPayInfoResponse(
                    request.callbackURL,
                    maxSendable,
                    minSendable,
                    "[[\"text/plain\",\"$plainTextMetadata\"]]",
                    "payRequest",
                )
            val success = replyServer(Json.encodeToString(response), request.replyURL)
            notifyChannel(
                context,
                NOTIFICATION_CHANNEL_LNURL_PAY,
                getString(
                    context,
                    if (success) LNURL_PAY_INFO_NOTIFICATION_TITLE else LNURL_PAY_NOTIFICATION_FAILURE_TITLE,
                    if (success) DEFAULT_LNURL_PAY_INFO_NOTIFICATION_TITLE else DEFAULT_LNURL_PAY_NOTIFICATION_FAILURE_TITLE
                ),
            )
        } catch (e: Exception) {
            logger.log(TAG, "Failed to process lnurl: ${e.message}", "WARN")
            if (request != null) {
                fail(e.message, request.replyURL)
            }
            notifyChannel(
                context,
                NOTIFICATION_CHANNEL_LNURL_PAY,
                getString(
                    context,
                    LNURL_PAY_NOTIFICATION_FAILURE_TITLE,
                    DEFAULT_LNURL_PAY_NOTIFICATION_FAILURE_TITLE
                ),
            )
        }

        fgService.shutdown()
    }
}

File: libs/sdk-bindings/bindings-swift/Sources/BreezSDK/Task/LnurlPayInfo.swift
import UserNotifications
import Foundation

struct LnurlInfoRequest: Codable {
    let callback_url: String
    let reply_url: String
}

struct LnurlInfoResponse: Decodable, Encodable {
    let callback: String
    let maxSendable: UInt64
    let minSendable: UInt64
    let metadata: String
    let tag: String
    
    init(callback: String, maxSendable: UInt64, minSendable: UInt64, metadata: String, tag: String) {
        self.callback = callback
        self.maxSendable = maxSendable
        self.minSendable = minSendable
        self.metadata = metadata
        self.tag = tag
    }
}

class LnurlPayInfoTask : LnurlPayTask {
    fileprivate let TAG = "LnurlPayInfoTask"
    
    init(payload: String, logger: ServiceLogger, config: ServiceConfig, contentHandler: ((UNNotificationContent) -> Void)? = nil, bestAttemptContent: UNMutableNotificationContent? = nil) {
        let successNotificationTitle = ResourceHelper.shared.getString(key: Constants.LNURL_PAY_INFO_NOTIFICATION_TITLE, fallback: Constants.DEFAULT_LNURL_PAY_INFO_NOTIFICATION_TITLE)
        let failNotificationTitle = ResourceHelper.shared.getString(key: Constants.LNURL_PAY_NOTIFICATION_FAILURE_TITLE, fallback: Constants.DEFAULT_LNURL_PAY_NOTIFICATION_FAILURE_TITLE)
        super.init(payload: payload, logger: logger, config: config, contentHandler: contentHandler, bestAttemptContent: bestAttemptContent, successNotificationTitle: successNotificationTitle, failNotificationTitle: failNotificationTitle)
    }
    
    override func start(breezSDK: BlockingBreezServices) throws {
        var request: LnurlInfoRequest? = nil
        do {
            request = try JSONDecoder().decode(LnurlInfoRequest.self, from: self.payload.data(using: .utf8)!)
        } catch let e {
            self.logger.log(tag: TAG, line: "failed to decode payload: \(e)", level: "ERROR")
            self.displayPushNotification(title: self.failNotificationTitle, logger: self.logger, threadIdentifier: Constants.NOTIFICATION_THREAD_LNURL_PAY)
            throw e
        }
        
        do {
            let plainTextMetadata = ResourceHelper.shared.getString(key: Constants.LNURL_PAY_METADATA_PLAIN_TEXT, fallback: Constants.DEFAULT_LNURL_PAY_METADATA_PLAIN_TEXT)
            let metadata = "[[\"text/plain\",\"\(plainTextMetadata)\"]]"
            // Get the fee parameters offered by the LSP for opening a new channel
            let ofp = try breezSDK.openChannelFee(req: OpenChannelFeeRequest(amountMsat: nil)).feeParams
            // Calculate the maximum receivable amount within the fee limit (in millisatoshis)
            let feeLimitMsat: UInt64 = config.channelFeeLimitMsat
            let nodeInfo = try breezSDK.nodeInfo()
            let proportionalPercent = Double(ofp.proportional) / 1000000.0
            // Treat fee limit feature as disabled when it's set to 0
            let maxReceivableMsatFeeLimit = (proportionalPercent != 0.0 && feeLimitMsat != 0)
            ? min(
                nodeInfo.maxReceivableMsat,
                UInt64(Double(feeLimitMsat) / proportionalPercent)
            ) : nodeInfo.maxReceivableMsat
            // Calculate the maximum sendable amount (in millisatoshis)
            let maxSendable = max(nodeInfo.inboundLiquidityMsats, maxReceivableMsatFeeLimit)
            // Get the minimum sendable amount (in millisatoshis), can not be less than 1 or more than maxSendable
            let minSendable: UInt64 = nodeInfo.inboundLiquidityMsats < UInt64(1000) ? ofp.minMsat :  UInt64(1000)
            if minSendable > maxSendable {
                throw InvalidMinSendable.largerThanMaxSendable
            }
            replyServer(encodable: LnurlInfoResponse(callback: request!.callback_url, maxSendable: maxSendable, minSendable: minSendable, metadata: metadata, tag: "payRequest"),
                        replyURL: request!.reply_url)
        } catch let e as InvalidMinSendable where e == .largerThanMaxSendable {
            self.logger.log(tag: TAG, line: "failed to process lnurl: \(e)", level: "ERROR")
            let failNotificationTitle = ResourceHelper.shared.getString(key: Constants.LNURL_PAY_NOTIFICATION_LIQUIDITY_FAILURE_TITLE, fallback: Constants.DEFAULT_LNURL_PAY_NOTIFICATION_LIQUIDITY_FAILURE_TITLE)
            fail(withError: e.localizedDescription, replyURL: request!.reply_url, failNotificationTitle: failNotificationTitle)
        } catch let e {
            self.logger.log(tag: TAG, line: "failed to process lnurl: \(e)", level: "ERROR")
            fail(withError: e.localizedDescription, replyURL: request!.reply_url)
        }
    }
}

private enum InvalidMinSendable: Error {
    case largerThanMaxSendable
}

extension InvalidMinSendable: LocalizedError {
    public var errorDescription: String? {
        switch self {
        case .largerThanMaxSendable:
            return NSLocalizedString("Minimum sendable amount can't be greater than maximum sendable amount.", comment: "InvalidMinSendable")
        }
    }
}
